packages/core/src/middleware.js

```diff
@@ -55,47 +65,46 @@ const applyMiddleware = (befores, afters, app, options) => {
   };
 
   return (input) => {
-    const context = ZapierPromise.makeContext();
-    const resolve = (val) => ZapierPromise.resolve(val).bind(context);
-
-    const beforeMiddleware = (beforeInput) => {
-      return befores.reduce((collector, func) => {
-        return collector.then((newInput) => {
-          newInput._addContext = context.addContext;
-          const args = [newInput].concat(options.extraArgs);
-          const result = func.apply(undefined, args);
-          if (typeof result !== 'object') {
-            throw new Error('Middleware should return an object.');
-          }
-          return result;
-        });
-      }, resolve(beforeInput));
+    const beforeMiddleware = async (beforeInput) => {
+      let newInput = beforeInput;
+      for (const func of befores) {
+        const args = [newInput].concat(options.extraArgs);
+        const maybePromise = func.apply(undefined, args);
+        // legacy scripting runner returns a Promise for beforeRequest
+        if (typeof maybePromise !== 'object') {
+          throw new Error('Middleware should return an object.');
+        }
+        newInput = await Promise.resolve(maybePromise);
```

Comment:
I added this promise logic here since the legacy scripting `beforeRequest` logic can now return a promise.